gateway/src/lib.rs

```diff
@@ -77,34 +138,109 @@ impl Server {
-    pub async fn wait_for_finish(self) -> Result<(), String> {
-        self.http_server.await
+    pub async fn wait_for_finish(&mut self) -> Result<(), String> {
+        while let Some(result) = self.all_servers_shutdown.next().await {
+            result?;
+        }
+        Ok(())
+    }
+
+    /// Adjust dropshot bind addresses on which this server is listening.
+    ///
+    /// Addresses fall into three categories:
+    ///
+    /// 1. Present in `addresses` and we're already listening on the address:
+    ///    the running dropshot server is left unchanged.
+    /// 2. Present in `addresses` but we're not already listening on the
+    ///    address: we start a new dropshot server (sharing the same server
+    ///    context) on the address.
+    /// 3. We're listening on it but it's not present in `addresses`: we shut
+    ///    the server down. This method blocks waiting for these shutdowns to
+    ///    complete.
+    ///
+    /// This method fails if any operation required for 2 or 3 fails.
+    pub async fn adjust_dropshot_addresses(
+        &mut self,
+        addresses: &[SocketAddrV6],
+    ) -> Result<(), String> {
+        if addresses.is_empty() {
+            return Err(
+                "Cannot reconfigure server with no addresses".to_string()
+            );
+        }
+
+        let mut http_servers = HashMap::with_capacity(addresses.len());
+
+        // For each address in `addresses`, either start a new server or move
+        // the existing one from `self.http_servers` into `http_servers`.
+        for &addr in addresses {
+            if let Some(existing) = self.http_servers.remove(&addr) {
+                info!(
+                    self.apictx.log,
+                    "adjusting dropshot addresses; keeping existing addr";
+                    "addr" => %addr,
+                );
+                http_servers.insert(addr, existing);
+            } else {
+                info!(
+                    self.apictx.log,
+                    "adjusting dropshot addresses; starting new server";
+                    "addr" => %addr,
+                );
+                start_dropshot_server(
+                    &self.apictx,
+                    addr,
+                    self.request_body_max_bytes,
+                    &mut http_servers,
+                    &self.all_servers_shutdown,
+                    &self.log,
+                )?;
+            }
+        }
+
+        // `http_servers` now contains all the servers we want for `addresses`.
+        // Swap it into `self.http_servers`, and shut down any servers that
+        // remain.
+        mem::swap(&mut http_servers, &mut self.http_servers);
+
+        for (addr, server) in http_servers {
+            info!(
+                self.apictx.log,
+                "adjusting dropshot addresses; stopping old server";
+                "addr" => %addr,
+            );
+            server.close().await?;
+        }
```

Comment:
Does this entry still exist in `all_servers_shutdown` when we do this, or does it get removed from that `FuturesUnordered` if someone is `await`-ing it via `wait_for_finish`?

Reply:
It does still exist in `all_servers_shutdown`. `wait_for_finish` loops until `all_servers_shutdown` is completely empty, though, so happens is:

0. Some task starts `await`ing `wait_for_finish`. Call this task A.
1. Task B calls this method to reconfigure servers.
2. Any new servers are started above this loop, and close handles for them are inserted into `all_servers_shutdown`. (Similarly, any existing servers are left running and left in `all_servers_shutdown`.)
3. Any servers being shut down are closed here.
4. Task A will see the "wait for finish" futures for the servers from step 3 pop out, but `wait_for_finish` the method still won't return, because the `FuturesUnordered` still contains the futures from step 2.

A caveat is that if any `server.close()` _fails_, the future in the `FuturesUnordered` will also fail with the same error, so `wait_for_finish` will return the error immediately. I think this is okay? If something goes wrong refreshing and we can't shut down a server, the safest thing to do is probably restart.

Reply:
Great explanation, thanks!